### PR TITLE
Support custom Lucene user data in engine commits

### PR DESCRIPTION
**Pull request by Tim-Brooks**
Various engine implementations may need to add additional user data to
the Lucene commit. This change allows extenders of InternalEngine to add
custom user-data.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 1]
```diff
@@ -2802,7 +2803,9 @@ protected void commitIndexWriter(final IndexWriter writer, final Translog transl
                  * {@link IndexWriter#commit()} call flushes all documents, we defer computation of the maximum sequence number to the time
                  * of invocation of the commit data iterator (which occurs after all documents have been flushed to Lucene).
                  */
-                final Map<String, String> commitData = Maps.newMapWithExpectedSize(8);
+                final Map<String, String> extraCommitUserData = getCommitExtraUserData();
+                final Map<String, String> commitData = Maps.newMapWithExpectedSize(8 + extraCommitUserData.size());
+                commitData.putAll(extraCommitUserData);
```
Comment (volodk85) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:2808]: Is there any upper bound for commit data? Any chances that we blow up that buffer if we allow potentially unlimited data in?

File: server/src/main/java/org/elasticsearch/TransportVersion.java [hunk 2]
```diff
@@ -144,9 +144,10 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
     public static final TransportVersion V_8_500_017 = registerTransportVersion(8_500_017, "0EDCB5BA-049C-443C-8AB1-5FA58FB996FB");
     public static final TransportVersion V_8_500_018 = registerTransportVersion(8_500_018, "827C32CE-33D9-4AC3-A773-8FB768F59EAF");
     public static final TransportVersion V_8_500_019 = registerTransportVersion(8_500_019, "09bae57f-cab8-423c-aab3-c9778509ffe3");
+    public static final TransportVersion V_8_500_020 = registerTransportVersion(8_500_020, "803e260d-9c0b-42a2-aa59-58e2b2024285");
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/TransportVersion.java:147]: FYI, this is needed for the stateless PR. The new getCommitExtraUserData() alone do not need a new transport version. Might be confusing for whoever in the future may try to find why this new transport version was introduced and may look at the commit/PR. I was definitely confused in the beginning of reading this PR why a new transport version was needed.

Do you think it'd make sense to do the new transport version in a separate PR? Or at least in the final commit description make a not that the transport version is unrelated to the new getCommitExtraUserData().

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (kingherc) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:2808]: I don't see an upper bound, but I'm also not sure I am concerned, as this is not user-provided, in the sense that it's code/plugins (extension of the InternalEngine) that will put any custom data. We should probably make sure ourselves we do not blow it up by putting unreasonable custom data. The limit seems to be memory or disk size. But I would not mind putting some sort of limit.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java [hunk 3]
```diff
@@ -2845,6 +2848,10 @@ protected void commitIndexWriter(final IndexWriter writer, final Translog transl
         }
     }
 
+    protected Map<String, String> getCommitExtraUserData() {
```
Comment (kingherc) [inline server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:2851]: A quick javadoc to explain that some keys may be reserved by ES and replaced, like ES_VERSION, may be useful.

**Review by kingherc — APPROVED** [review]
LGTM. But have a few comments that may warrant some attention.

File: server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java (hunk 1 above)
Comment (Tim-Brooks) [reply server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java:2808]: This is set by extenders of the `InternalEngine`. Adding user-data tends to be a deliberate decision and only occasionally added. Obviously a poorly implemented Engine implementation could add excessive user-data which could cause issues. But it is an Engine implementation, it has full access to everything (memory, disk, etc) so it generally trusted.

**Pull request merged**